packages/zero-cache/src/lock.ts

```diff
@@ -0,0 +1,96 @@
+import type {LogContext} from '@rocicorp/logger';
+import {resolver} from '@rocicorp/resolver';
+
+let id = 0;
+export class Lock {
+  #lockP: {promise: Promise<void>; name: string} | null = null;
+  readonly #lc: LogContext | undefined;
+  readonly #id: number;
+  readonly #pollInterval: number;
+
+  constructor(lc?: LogContext | undefined, pollInterval: number = 1000) {
+    this.#id = id++;
+    this.#pollInterval = pollInterval;
+    this.#lc = lc
+      ?.withContext('component', 'lock')
+      ?.withContext('id', this.#id);
+  }
+
+  async lock(taskName: string): Promise<() => void> {
+    const previous = this.#lockP;
+    const {promise, resolve} = resolver();
+    this.#lockP = {promise, name: taskName};
+
+    let waiting = true;
+    const acquisitionStack = new Error('Lock acquisition stack');
```

Comment:
nit: we should only acquire `acquisitionStack` if the log level is set to debug.